packages/backend/src/services/action.ts

```diff
@@ -46,6 +93,29 @@ export const processAction = async (options: ProcessActionOptions) => {
   )
 
   $.step.parameters = computedParameters
+  if (rateLimiters[$.app.key] && worker) {
+    const { isRateLimited, duration } = await rateLimiters[$.app.key](
+      $.step.parameters,
+    )
+    if (isRateLimited) {
+      await actionQueue.add(
```

Comment:
Hmmm.... This will re-queue the exact same action, but how will we handle a case where a user tries to send > `$rate_limit` emails via postman? (e.g. `destination` array > 169 people).

Will the 1st 169 people keep getting emailed repeatedly?

Wondering if we can magically ensure that we can always make progress each time we re-queue.

Reply:
One way i can think of is to introduce a new `context` column to `ExecutionStep`, and use that to store which emails we've processed already. So each time action gets re-queued, we just exclude the emails we've already sent.

(I'm kinda planning on using a similar `context` column to implement for-each, you can read about it in if-then design doc final section)

_Small aside_
Earlier on, I was evaluating [`rate-limiter-flexible`](https://www.npmjs.com/package/rate-limiter-flexible) vs [`bottleneck`](https://www.npmjs.com/package/bottleneck). The latter has "_enqueue all callbacks and execute at X QPS_" functionality.

I decided on the latter because I didn't think we need queueing capability, but one easy way out is to migrate to `bottleneck`. Then we can throw all the postman API calls into the `bottleneck` queue and just ask it to execute at 169 QPS. But I really don't like this because it means our worker is going to idle between each call :(

Reply:
also thinking if we should resolve this rate limit problem in separate PR, so that we can land `noreply@plumber.gov.sg` function and let users avoid the "get postman API key" step.